File: zena_mode/handlers/voice.py
import asyncio
import logging
from zena_mode.handlers.base import BaseZenHandler
from zena_mode.voice_manager import get_voice_manager
from config_system import config

logger = logging.getLogger("ZenAI.Handler.Voice")


class VoiceHandler:
    """Namespace for voice-related request handling."""

    @staticmethod
    def handle_get(handler: BaseZenHandler):
        """Routing for GET requests related to voice."""
        path = handler.path

        # GET /voice/devices - Enumerate microphones
        if path == "/voice/devices":
            try:
                vm = get_voice_manager()
                devices = vm.enumerate_devices()
                result = {
                    "success": True,
                    "devices": [d.to_dict() for d in devices],
                    "default_device": vm.get_default_input_device(),
                }
                handler.send_json_response(200, result)
            except Exception as e:
                logger.error(f"Device enumeration failed: {e}")
                handler.send_json_response(500, {"success": False, "error": str(e)})
            return True

        # GET /voice/status - Get voice system status
        elif path == "/voice/status":
            try:
                vm = get_voice_manager()
                status = vm.get_status()
                handler.send_json_response(200, status)
            except Exception as e:
                logger.error(f"Status check failed: {e}")
                handler.send_json_response(500, {"success": False, "error": str(e)})
            return True

        # GET /voice/lab - Serve voice lab HTML (legacy)
        elif path == "/voice/lab":
            try:
                lab_path = config.BASE_DIR / "static" / "voice_lab.html"
                if lab_path.exists():
                    handler.send_response(200)
                    handler.send_header("Content-Type", "text/html")
                    handler.end_headers()
                    with open(lab_path, "rb") as f:
                        handler.wfile.write(f.read())
                else:
                    handler.send_json_response(404, {"error": "Voice Lab UI missing"})
            except Exception as e:
                handler.send_json_response(500, {"error": str(e)})
            return True

        # GET /api/tts-voices - List available TTS voices (edge-tts neural)
        elif path == "/api/tts-voices":
            try:
                from Core.tts_engine import TTSEngine
                loop = asyncio.new_event_loop()
                try:
                    voices = loop.run_until_complete(TTSEngine.list_voices())
                finally:
                    loop.close()
                result = {
                    "voices": [
                        {"id": v.get("name", ""), "name": v.get("name", ""), "gender": v.get("gender", "Unknown")}
                        for v in voices
                    ]
                }
                handler.send_json_response(200, result)
            except Exception as e:
                logger.error(f"TTS voices listing failed: {e}")
                handler.send_json_response(500, {"error": str(e), "voices": []})
            return True

        return False

    @staticmethod
    def handle_post(handler: BaseZenHandler):
        """Routing for POST requests related to voice."""
        path = handler.path

        # POST /api/test-audio - Play a test tone on the selected speaker
        if path == "/api/test-audio":
            try:
                import sounddevice as sd
                from zena_mode.production_microphone_healer import MicrophoneHealer

                params = handler.parse_json_body()
                device_id = params.get("device_id")
                if device_id is not None:
                    device_id = int(device_id)

                healer = MicrophoneHealer()
                tone_bytes = healer.generate_test_tone(frequency=1000, duration=0.3)

                import io
                from scipy.io import wavfile

                wav_buf = io.BytesIO(tone_bytes)
                rate, audio = wavfile.read(wav_buf)
                audio_float = audio.astype("float32") / 32768.0

                sd.play(audio_float, samplerate=rate, device=device_id)
                sd.wait()
                handler.send_json_response(200, {"status": "ok", "msg": "Test tone played"})
            except Exception as e:
                logger.error(f"test-audio failed: {e}")
                handler.send_json_response(500, {"status": "error", "msg": str(e)})
            return True

        # POST /api/test-loopback - Play tone and listen for it
        elif path == "/api/test-loopback":
            try:
                from zena_mode.production_microphone_healer import MicrophoneHealer

                params = handler.parse_json_body()
                input_id = params.get("input_id")
                output_id = params.get("output_id")
                if input_id is not None:
                    input_id = int(input_id)
                if output_id is not None:
                    output_id = int(output_id)

                device_id = input_id if input_id is not None else (output_id or 0)
                healer = MicrophoneHealer()
                success, confidence, reason = healer.verify_loopback(device_id)
                handler.send_json_response(200, {
                    "status": "ok",
                    "success": success,
                    "magnitude": confidence * 100,
                    "msg": reason,
                })
            except Exception as e:
                logger.error(f"test-loopback failed: {e}")
                handler.send_json_response(500, {"status": "error", "msg": str(e)})
            return True

        # POST /api/record - Record audio and transcribe
        if path == "/api/record":
            try:
                params = handler.parse_json_body()
                device_id = params.get("device_id")
                duration = params.get("duration", 3.0)

                if device_id is not None:
                    device_id = int(device_id)

                # Record audio
                vm = get_voice_manager()
                recording_result = vm.record_audio(duration=duration, device_id=device_id)

                if not recording_result.success:
                    handler.send_json_response(500, {"success": False, "error": recording_result.error})
                    return True

                # Transcribe
                transcribe_result = vm.transcribe(recording_result.audio_data)
                result = {
                    "success": transcribe_result.get("success", False),
                    "text": transcribe_result.get("text", ""),
                    "duration": recording_result.duration,
                    "error": transcribe_result.get("error"),
                }
                handler.send_json_response(200, result)
            except Exception as e:
                logger.error(f"Record endpoint failed: {e}")
                handler.send_json_response(500, {"success": False, "error": str(e)})
            return True

        # POST /voice/transcribe - Transcribe uploaded audio
        elif path == "/voice/transcribe":
            try:
                content_length = int(handler.headers.get("Content-Length", 0))
                audio_data = handler.rfile.read(content_length)

                vm = get_voice_manager()
                result = vm.transcribe(audio_data)
                handler.send_json_response(200, result)
            except Exception as e:
                logger.error(f"Transcribe endpoint failed: {e}")
                handler.send_json_response(500, {"success": False, "error": str(e)})
            return True

        # POST /voice/synthesize - Generate speech from text
        elif path == "/voice/synthesize":
            try:
                params = handler.parse_json_body()
                text = params.get("text", "")

                if not text:
                    handler.send_json_response(400, {"success": False, "error": "Missing 'text' parameter"})
                    return True

                vm = get_voice_manager()
                result = vm.synthesize(text)
                handler.send_json_response(200, result)
            except Exception as e:
                logger.error(f"Synthesize endpoint failed: {e}")
                handler.send_json_response(500, {"success": False, "error": str(e)})
            return True

        # POST /voice/speak - Alias for synthesize (more intuitive name)
        elif path == "/voice/speak":
            try:
                params = handler.parse_json_body()
                text = params.get("text", "")

                if not text:
                    handler.send_json_response(400, {"success": False, "error": "Missing 'text' parameter"})
                    return True

                vm = get_voice_manager()
                result = vm.synthesize(text, use_cache=True)

                # Return just the audio URL for easy HTML5 integration
                handler.send_json_response(
                    200,
                    {
                        "success": result["success"],
                        "audio_url": result.get("audio_url"),
                        "duration": result.get("duration"),
                        "text": text,
                        "error": result.get("error"),
                    },
                )
            except Exception as e:
                logger.error(f"Speak endpoint failed: {e}")
                handler.send_json_response(500, {"success": False, "error": str(e)})
            return True

        return False

File: rust_output/src/voice.rs
use anyhow::{Result, Context};
use crate::base::{BaseZenHandler};
use crate::config_system::{config};
use crate::voice_manager::{get_voice_manager};
use std::collections::HashMap;
use std::fs::File;
use std::io::{self, Read, Write};

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

/// Namespace for voice-related request handling.
#[derive(Debug, Clone)]
pub struct VoiceHandler {
}

impl VoiceHandler {
    /// Routing for GET requests related to voice.
    pub fn handle_get(handler: BaseZenHandler) -> Result<()> {
        // Routing for GET requests related to voice.
        let mut path = handler.path;
        if path == "/voice/devices".to_string() {
            // try:
            {
                let mut vm = get_voice_manager();
                let mut devices = vm.enumerate_devices();
                let mut result = HashMap::from([("success".to_string(), true), ("devices".to_string(), devices.iter().map(|d| d.to_dict()).collect::<Vec<_>>()), ("default_device".to_string(), vm.get_default_input_device())]);
                handler.send_json_response(200, result);
            }
            // except Exception as e:
            true
        } else if path == "/voice/status".to_string() {
            // try:
            {
                let mut vm = get_voice_manager();
                let mut status = vm.get_status();
                handler.send_json_response(200, status);
            }
            // except Exception as e:
            true
        } else if path == "/voice/lab".to_string() {
            // try:
            {
                let mut lab_path = ((config::BASE_DIR / "static".to_string()) / "voice_lab.html".to_string());
                if lab_path.exists() {
                    handler.send_response(200);
                    handler.send_header("Content-Type".to_string(), "text/html".to_string());
                    handler.end_headers();
                    let mut f = File::open(lab_path)?;
                    {
                        handler.wfile.write(f.read());
                    }
                } else {
                    handler.send_json_response(404, HashMap::from([("error".to_string(), "Voice Lab UI missing".to_string())]));
                }
            }
            // except Exception as e:
            true
        } else if path == "/api/tts-voices".to_string() {
            // try:
            {
                // TODO: from Core.tts_engine import TTSEngine
                let mut r#loop = asyncio.new_event_loop();
                // try:
                {
                    let mut voices = r#loop.run_until_complete(TTSEngine.list_voices());
                }
                // finally:
                    r#loop.close();
                let mut result = HashMap::from([("voices".to_string(), voices.iter().map(|v| HashMap::from([("id".to_string(), v.get(&"name".to_string()).cloned().unwrap_or("".to_string())), ("name".to_string(), v.get(&"name".to_string()).cloned().unwrap_or("".to_string())), ("gender".to_string(), v.get(&"gender".to_string()).cloned().unwrap_or("Unknown".to_string()))])).collect::<Vec<_>>())]);
                handler.send_json_response(200, result);
            }
            // except Exception as e:
            true
        }
        Ok(false)
    }
    /// Routing for POST requests related to voice.
    pub fn handle_post(handler: BaseZenHandler) -> Result<()> {
        // Routing for POST requests related to voice.
        let mut path = handler.path;
        if path == "/api/test-audio".to_string() {
            // try:
            {
                // TODO: import sounddevice as sd
                // TODO: from zena_mode.production_microphone_healer import MicrophoneHealer
                let mut params = handler.parse_json_body();
                let mut device_id = params.get(&"device_id".to_string()).cloned();
                if device_id.is_some() {
                    let mut device_id = device_id.to_string().parse::<i64>().unwrap_or(0);
                }
                let mut healer = MicrophoneHealer();
                let mut tone_bytes = healer.generate_test_tone(/* frequency= */ 1000, /* duration= */ 0.3_f64);
                // TODO: import io
                // TODO: from scipy.io import wavfile
                let mut wav_buf = io.BytesIO(tone_bytes);
                let (mut rate, mut audio) = wavfile.read(wav_buf);
                let mut audio_float = (audio.astype("float32".to_string()) / 32768.0_f64);
                sd.play(audio_float, /* samplerate= */ rate, /* device= */ device_id);
                sd.wait();
                handler.send_json_response(200, HashMap::from([("status".to_string(), "ok".to_string()), ("msg".to_string(), "Test tone played".to_string())]));
            }
            // except Exception as e:
            true
        } else if path == "/api/test-loopback".to_string() {
            // try:
            {
                // TODO: from zena_mode.production_microphone_healer import MicrophoneHealer
                let mut params = handler.parse_json_body();
                let mut input_id = params.get(&"input_id".to_string()).cloned();
                let mut output_id = params.get(&"output_id".to_string()).cloned();
                if input_id.is_some() {
                    let mut input_id = input_id.to_string().parse::<i64>().unwrap_or(0);
                }
                if output_id.is_some() {
                    let mut output_id = output_id.to_string().parse::<i64>().unwrap_or(0);
                }
                let mut device_id = if input_id.is_some() { input_id } else { (output_id || 0) };
                let mut healer = MicrophoneHealer();
                let (mut success, mut confidence, mut reason) = healer.verify_loopback(device_id);
                handler.send_json_response(200, HashMap::from([("status".to_string(), "ok".to_string()), ("success".to_string(), success), ("magnitude".to_string(), (confidence * 100)), ("msg".to_string(), reason)]));
            }
            // except Exception as e:
            true
        }
        if path == "/api/record".to_string() {
            // try:
            {
                let mut params = handler.parse_json_body();
                let mut device_id = params.get(&"device_id".to_string()).cloned();
                let mut duration = params.get(&"duration".to_string()).cloned().unwrap_or(3.0_f64);
                if device_id.is_some() {
                    let mut device_id = device_id.to_string().parse::<i64>().unwrap_or(0);
                }
                let mut vm = get_voice_manager();
                let mut recording_result = vm.record_audio(/* duration= */ duration, /* device_id= */ device_id);
                if !recording_result.success {
                    handler.send_json_response(500, HashMap::from([("success".to_string(), false), ("error".to_string(), recording_result.error)]));
                    true
                }
                let mut transcribe_result = vm.transcribe(recording_result.audio_data);
                let mut result = HashMap::from([("success".to_string(), transcribe_result.get(&"success".to_string()).cloned().unwrap_or(false)), ("text".to_string(), transcribe_result.get(&"text".to_string()).cloned().unwrap_or("".to_string())), ("duration".to_string(), recording_result.duration), ("error".to_string(), transcribe_result.get(&"error".to_string()).cloned())]);
                handler.send_json_response(200, result);
            }
            // except Exception as e:
            true
        } else if path == "/voice/transcribe".to_string() {
            // try:
            {
                let mut content_length = handler.headers.get(&"Content-Length".to_string()).cloned().unwrap_or(0).to_string().parse::<i64>().unwrap_or(0);
                let mut audio_data = handler.rfile.read(content_length);
                let mut vm = get_voice_manager();
                let mut result = vm.transcribe(audio_data);
                handler.send_json_response(200, result);
            }
            // except Exception as e:
            true
        } else if path == "/voice/synthesize".to_string() {
            // try:
            {
                let mut params = handler.parse_json_body();
                let mut text = params.get(&"text".to_string()).cloned().unwrap_or("".to_string());
                if !text {
                    handler.send_json_response(400, HashMap::from([("success".to_string(), false), ("error".to_string(), "Missing 'text' parameter".to_string())]));
                    true
                }
                let mut vm = get_voice_manager();
                let mut result = vm.synthesize(text);
                handler.send_json_response(200, result);
            }
            // except Exception as e:
            true
        } else if path == "/voice/speak".to_string() {
            // try:
            {
                let mut params = handler.parse_json_body();
                let mut text = params.get(&"text".to_string()).cloned().unwrap_or("".to_string());
                if !text {
                    handler.send_json_response(400, HashMap::from([("success".to_string(), false), ("error".to_string(), "Missing 'text' parameter".to_string())]));
                    true
                }
                let mut vm = get_voice_manager();
                let mut result = vm.synthesize(text, /* use_cache= */ true);
                handler.send_json_response(200, HashMap::from([("success".to_string(), result["success".to_string()]), ("audio_url".to_string(), result.get(&"audio_url".to_string()).cloned()), ("duration".to_string(), result.get(&"duration".to_string()).cloned()), ("text".to_string(), text), ("error".to_string(), result.get(&"error".to_string()).cloned())]));
            }
            // except Exception as e:
            true
        }
        Ok(false)
    }
}
